File: benchmarks/python/higher_order.py
from collections.abc import Callable


def double(x: int) -> int:
    return x * 2


def triple(x: int) -> int:
    return x * 3


def apply_twice(f: Callable[[int], int], x: int) -> int:
    return f(f(x))


def describe(f: Callable[[int], int], label: str) -> None:
    print(f"{label} says {f(5)}")


def main() -> None:
    d = double
    print(f"double twice: {apply_twice(d, 5)}")
    print(f"triple twice: {apply_twice(triple, 5)}")
    describe(d, "double")
    describe(triple, "triple")


if __name__ == "__main__":
    main()

File: benchmarks/rust/higher_order.rs
fn double(x: i64) -> i64 {
    x * 2
}

fn triple(x: i64) -> i64 {
    x * 3
}

fn apply_twice(f: fn(i64) -> i64, x: i64) -> i64 {
    f(f(x))
}

fn describe(f: fn(i64) -> i64, label: &str) {
    println!("{} says {}", label, f(5));
}

fn main() {
    let d = double as fn(i64) -> i64;
    println!("double twice: {}", apply_twice(d, 5));
    println!("triple twice: {}", apply_twice(triple, 5));
    describe(d, "double");
    describe(triple, "triple");
}
